torch/csrc/jit/passes/automatic_mixed_precision.cpp

```diff
@@ -0,0 +1,263 @@
+          scope.instance = value;
+          scope.enabled = def->i(attr::value) ? true : false;
+          return scope;
+        }
+      }
+    }
+  }
+
+  // Not an autocast...
+  return c10::nullopt;
+}
+
+void castTensorInputs(Node* node, at::ScalarType dtype) {
+  const auto graph = node->owningGraph();
+
+  WithInsertPoint insert_point(node);
+
+  const auto dtype_value = graph->insertConstant(dtype);
+  const auto false_value = graph->insertConstant(false);
+  const auto none_value = graph->insertConstant(IValue());
+
+  std::unordered_set<Value*> casted_inputs;
+
+  for (auto input : node->inputs()) {
+    if (input->type()->kind() == TensorType::Kind) {
+      casted_inputs.insert(input);
+    }
+  }
+
+  for (auto input : casted_inputs) {
+    const auto new_input = graph->insert(
+        aten::to, {input, dtype_value, false_value, false_value, none_value});
+    node->replaceInputWith(input, new_input);
+  }
+}
+
+void castInputsToWidestType(Node* node) {
+  // Figure out the widest type
+  // (really, just looking for any float32 inputs)
+  //
+  // TODO: revisit this (do we need to consider float64 types?)
+  //
+  for (auto input : node->inputs()) {
+    if (auto tensor_type = input->type()->cast<TensorType>()) {
+      const auto dtype = tensor_type->scalarType();
+      if (!dtype.has_value() || *dtype != at::ScalarType::Half) {
+        castTensorInputs(node, at::ScalarType::Float);
+        return;
+      }
+    }
+  }
+}
+
+void handleBlock(Block* block, bool initial_state) {
+  std::stack<AutocastScope> autocast_stack;
+
+  // The current autocast enabled/disabled state
+  auto current_state = [&] {
+    return autocast_stack.empty() ? initial_state
+                                  : autocast_stack.top().enabled;
+  };
+
+  for (Node* node : block->nodes()) {
+    switch (node->kind()) {
+      case prim::CallFunction:
+      case prim::CallMethod:
+        TORCH_INTERNAL_ASSERT(false, "Calls are not expected with AMP & JIT");
+        break;
+
+      case prim::Enter:
+        if (auto autocast_scope = parseAutocast(node->input())) {
+          autocast_stack.push(*autocast_scope);
+        }
+        break;
+
+      case prim::Exit:
+        // TODO: technically we can avoid parseAutocast() here
+        if (auto autocast_scope = parseAutocast(node->input())) {
+          TORCH_INTERNAL_ASSERT(!autocast_stack.empty());
+          TORCH_INTERNAL_ASSERT(
+              autocast_stack.top().instance == autocast_scope->instance);
+          autocast_stack.pop();
+        }
+        break;
+
+      // CastPolicy::fp16 (cast all inputs to float16)
+      case aten::_convolution:
+      case aten::_convolution_nogroup:
+      case aten::conv1d:
+      case aten::conv2d:
+      case aten::conv3d:
+      case aten::conv_tbc:
+      case aten::conv_transpose1d:
+      case aten::convolution:
+      case aten::cudnn_convolution:
+      case aten::cudnn_convolution_transpose:
+      case aten::prelu:
```

Comment:
Is this right? @mcarilli

Reply:
Yes, this mimics the eager cast list. I don't remember why prelu was on the fp16 list, but it doesn't seem to do any harm.  maybe we should remove it, but that seems low priority.  Before we start updating the cast list, I'd rather get to a state where the list is hardcoded in just one place that jit and eager both draw from.  Making changes in two places is brittle.

Reply:
I agree re. unifying the policy definitions between eager and jit. It's on the TODO list.

My initial attempt looked too messy so for the proof of concept I manually duplicated the definitions. One shortcut in this interim state is that I don't differentiate between overloads. @mcarilli , are there any specific cases where different overloads have different policies?